### ISSUE TITLE
Bindings invoke disposed ViewModels even though the DataContext was set to null beforehand

### ISSUE THREAD

**Issue by jeanplevesque** [unoplatform/uno#11956]
### Context

Our client complains about (non-fatal) exceptions (thrown from ViewModels) being reported in their diagnostics dashboard.
After some investigation and some attempted fixes, I was able to resolve the issues for the UWP version of the app. However, the mobile version still have the problem.

### Current behavior

Basically, the `Uno.UI.DataBinding.BindingPropertyHelper` invokes properties on some disposed ViewModels, which leads to exceptions.

We tried to fix this by ensuring that we **set the `DataContext` of pages to `null` before disposing it** (the ViewModel).
This seems to work fine on UWP, but not on mobile.

### Expected behavior

Mobile behaves like Windows.

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

4.6.44

### Affected platforms

Android, iOS

### IDE

Visual Studio 2022

### IDE version

17.3.6

**Comment by jeromelaban** [comment]
Thanks @jeanplevesque. Can you provide a repro?

**Comment by jeanplevesque** [comment]
@jeromelaban I tried with Uno.UI 4.9.0-dev.539 and I get the same behavior.

**Comment by jeromelaban** [comment]
Thanks. We'll still need a repro or any kind of information that could show what is causing the bindings to still update after it's been removed from its target.

**Comment by jeanplevesque** [comment]
I'm having trouble creating a sample that reproduces the problem.
It seems more pieces are involved. My current hypotheses are around `DataTemplate`, `DataContext` inheritance, and binding path length.

**Comment by jeromelaban** [comment]
> @jeromelaban Is that possible? Is there some code (optimizations perhaps) that could cache or memoize things around the binding system?

There is some caching of the getters/setters, but not of the values.

**Comment by jeanplevesque** [comment]
@jeromelaban Here's the call stack when my exception occurs in my disposed ViewModel.

```
 	0x7 in ViewModel.GetPropertyValue at C:\Repos\App\src\app\ViewModel.cs:105,4	C#
 	0xE in ViewModel.get_Property at C:\Repos\App\src\app\ViewModel.cs:99,11	C#
 	0xFFFFFFFFFFFFFFFF in System.Reflection.RuntimeMethodInfo.InternalInvoke	C#
 	0x70 in System.Reflection.RuntimeMethodInfo.Invoke at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/corlib/System.Reflection/RuntimeMethodInfo.cs:395,6	C#
 	0x6 in System.Reflection.MethodBase.Invoke at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/external/corefx/src/Common/src/CoreLib/System/Reflection/MethodBase.cs:53,66	C#
 	0x8 in Uno.UI.DataBinding.BindingPropertyHelper.	C#
 	0xC in Uno.UI.DataBinding.BindingPropertyHelper.	C#
 	0xC in Uno.UI.DataBinding.BindingPath.BindingItem.GetSourceValue	C#
 	0xD in Uno.UI.DataBinding.BindingPath.BindingItem.GetPrecedenceSpecificValue	C#
 	0x17 in Uno.UI.DataBinding.BindingPath.set_Value	C#
 	0x81 in Windows.UI.Xaml.Data.BindingExpression.UpdateSource	C#
 	0x23 in Windows.UI.Xaml.Data.BindingExpression.SetSourceValue	C#
 	0xC in Windows.UI.Xaml.DependencyPropertyDetails.SetSourceValue	C#
 	0x2 in Windows.UI.Xaml.DependencyPropertyDetailsCollection.SetSourceValue	C#
 	0x3C in Windows.UI.Xaml.DependencyObjectStore.SetBindingValue	C#
 	0x8 in Windows.UI.Xaml.DependencyObjectStore.OnDependencyPropertyChanged	C#
 	0x102 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x6 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x8 in Uno.UI.Controls.BindableView.SetValue	C#
 	0x7 in Windows.UI.Xaml.Controls.Primitives.Selector.set_SelectedItem	C#
 	0x2F in Windows.UI.Xaml.Controls.Primitives.Selector.OnSelectedIndexChanged	C#
 	0x4C in Windows.UI.Xaml.Controls.Primitives.Selector.OnPropertyChanged2	C#
 	0x2 in Uno.UI.Controls.BindableView.Windows.UI.Xaml.IDependencyObjectInternal.OnPropertyChanged2	C#
 	0x116 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x6 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x8 in Uno.UI.Controls.BindableView.SetValue	C#
 	0xC in Windows.UI.Xaml.Controls.Primitives.Selector.set_SelectedIndex	C#
 	0xB in Windows.UI.Xaml.Controls.Primitives.Selector.ResetIndexIfNeeded	C#
 	0x159 in Windows.UI.Xaml.Controls.Primitives.Selector.OnItemsSourceSingleCollectionChanged	C#
 	0x21 in Windows.UI.Xaml.Controls.ItemsControl.OnItemsVectorChanged	C#
 	0x35 in Uno.UI.Extensions.VectorChangedEventHandlerExtensions.TryRaiseReseted	C#
 	0x73 in Windows.UI.Xaml.Controls.ItemCollection.SetItemsSource	C#
 	0x6B in Windows.UI.Xaml.Controls.ItemsControl.OnItemsSourceChanged	C#
 	0x2 in Windows.UI.Xaml.Controls.Primitives.Selector.OnItemsSourceChanged	C#
 	0x7 in Windows.UI.Xaml.Controls.ItemsControl.<>c.<.cctor>b__213_4	C#
 	0xD in Windows.UI.Xaml.PropertyMetadata.RaisePropertyChanged	C#
 	0xFA in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x1F in Windows.UI.Xaml.DependencyObjectExtensions.SetValue	C#
 	0x79 in Uno.UI.DataBinding.BindingPropertyHelper.	C#
 	0x17 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValue	C#
 	0x77 in Windows.UI.Xaml.Data.BindingExpression.ApplyFallbackValue	C#
 	0xA in Windows.UI.Xaml.Data.BindingExpression.InnerSetTargetValueSafe	C#
 	0x3 in Windows.UI.Xaml.Data.BindingExpression.<SetTargetValueSafe>g__SetTargetValueSafeWithTry|54_1	C#
 	0xC0 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x19 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x16 in Windows.UI.Xaml.Data.BindingExpression.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.BindingItem.RaiseValueChanged	C#
 	0x50 in Uno.UI.DataBinding.BindingPath.BindingItem.OnDataContextChanged	C#
 	0xF in Uno.UI.DataBinding.BindingPath.BindingItem.SetWeakDataContext	C#
 	0x15 in Uno.UI.DataBinding.BindingPath.OnDataContextChanged	C#
 	0x17 in Uno.UI.DataBinding.BindingPath.SetWeakDataContext	C#
 	0xA3 in Windows.UI.Xaml.Data.BindingExpression.ApplyBinding	C#
 	0x3F in Windows.UI.Xaml.Data.BindingExpression.set_DataContext	C#
 	0x6B in Windows.UI.Xaml.DependencyPropertyDetailsCollection.ApplyBinding	C#
 	0x15 in Windows.UI.Xaml.DependencyPropertyDetailsCollection.ApplyDataContext	C#
 	0x7 in Windows.UI.Xaml.DependencyObjectStore.ApplyDataContext	C#
 	0x58 in Windows.UI.Xaml.DependencyObjectStore.OnDataContextChanged	C#
 	0xDA in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0xA in Windows.UI.Xaml.DependencyObjectStore.ClearValue	C#
 	0x30 in Windows.UI.Xaml.Controls.ContentControl.ResetContentDataContextOverride	C#
 	0xB in Windows.UI.Xaml.Controls.ContentControl.set_ContentTemplateRoot	C#
 	0x2E in Windows.UI.Xaml.Controls.ContentControl.OnContentChanged	C#
 	0x17 in Windows.UI.Xaml.Controls.ContentControl.<>c.<.cctor>b__65_0	C#
 	0xD in Windows.UI.Xaml.PropertyMetadata.RaisePropertyChanged	C#
 	0xFA in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x1F in Windows.UI.Xaml.DependencyObjectExtensions.SetValue	C#
 	0x79 in Uno.UI.DataBinding.BindingPropertyHelper.	C#
 	0x17 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValue	C#
 	0x5C in Windows.UI.Xaml.Data.BindingExpression.InnerSetTargetValueSafe	C#
 	0x3 in Windows.UI.Xaml.Data.BindingExpression.<SetTargetValueSafe>g__SetTargetValueSafeWithTry|54_1	C#
 	0xC0 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x19 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x16 in Windows.UI.Xaml.Data.BindingExpression.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.BindingItem.RaiseValueChanged	C#
 	0x85 in Uno.UI.DataBinding.BindingPath.BindingItem.OnDataContextChanged	C#
 	0xF in Uno.UI.DataBinding.BindingPath.BindingItem.SetWeakDataContext	C#
 	0x29 in Uno.UI.DataBinding.BindingPath.BindingItem.set_DataContext	C#
 	0x69 in Uno.UI.DataBinding.BindingPath.BindingItem.OnDataContextChanged	C#
 	0xF in Uno.UI.DataBinding.BindingPath.BindingItem.SetWeakDataContext	C#
 	0x15 in Uno.UI.DataBinding.BindingPath.OnDataContextChanged	C#
 	0x17 in Uno.UI.DataBinding.BindingPath.SetWeakDataContext	C#
 	0xA3 in Windows.UI.Xaml.Data.BindingExpression.ApplyBinding	C#
 	0x3F in Windows.UI.Xaml.Data.BindingExpression.set_DataContext	C#
 	0x6B in Windows.UI.Xaml.DependencyPropertyDetailsCollection.ApplyBinding	C#
 	0x15 in Windows.UI.Xaml.DependencyPropertyDetailsCollection.ApplyDataContext	C#
 	0x7 in Windows.UI.Xaml.DependencyObjectStore.ApplyDataContext	C#
 	0x58 in Windows.UI.Xaml.DependencyObjectStore.OnDataContextChanged	C#
 	0xDA in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x4A in Windows.UI.Xaml.DependencyObjectStore.OnParentPropertyChangedCallback	C#
 	0x31 in Windows.UI.Xaml.DependencyObjectStore.CallChildCallback	C#
 	0xC0 in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x6 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x8 in Uno.UI.Controls.BindableView.SetValue	C#
 	0x7 in Uno.UI.Controls.BindableView.set_DataContext	C#
 	0xD in Windows.UI.Xaml.Controls.ContentPresenter.TrySetDataContextFromContent	C#
 	0x2B in Windows.UI.Xaml.Controls.ContentPresenter.OnContentChanged	C#
 	0x17 in Windows.UI.Xaml.Controls.ContentPresenter.<>c.<.cctor>b__277_0	C#
 	0xD in Windows.UI.Xaml.PropertyMetadata.RaisePropertyChanged	C#
 	0xFA in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x1F in Windows.UI.Xaml.DependencyObjectExtensions.SetValue	C#
 	0x79 in Uno.UI.DataBinding.BindingPropertyHelper.	C#
 	0x17 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValue	C#
 	0x5C in Windows.UI.Xaml.Data.BindingExpression.InnerSetTargetValueSafe	C#
 	0x3 in Windows.UI.Xaml.Data.BindingExpression.<SetTargetValueSafe>g__SetTargetValueSafeWithTry|54_1	C#
 	0xC0 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x19 in Windows.UI.Xaml.Data.BindingExpression.SetTargetValueSafe	C#
 	0x16 in Windows.UI.Xaml.Data.BindingExpression.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.Uno.UI.DataBinding.IValueChangedListener.OnValueChanged	C#
 	0xC in Uno.UI.DataBinding.BindingPath.BindingItem.RaiseValueChanged	C#
 	0x2E in Uno.UI.DataBinding.BindingPath.BindingItem.OnPropertyChanged	C#
 	0x1A in Uno.UI.DataBinding.BindingPath.BindingItem.PropertyChangedValueHandler.NewValue	C#
 	0x1 in Uno.UI.DataBinding.BindingPath.BindingItem.PropertyChangedValueHandler.NewValue	C#
 	0x27 in Windows.UI.Xaml.DependencyObjectStore.	C#
 	0x30 in Windows.UI.Xaml.DependencyPropertyCallbackManager.RaisePropertyChanged	C#
 	0xD in Windows.UI.Xaml.DependencyPropertyDetails.RaisePropertyChanged	C#
 	0x11E in Windows.UI.Xaml.DependencyObjectStore.InvokeCallbacks	C#
 	0x6C in Windows.UI.Xaml.DependencyObjectStore.RaiseCallbacks	C#
 	0x1D4 in Windows.UI.Xaml.DependencyObjectStore.InnerSetValue	C#
 	0x22 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x6 in Windows.UI.Xaml.DependencyObjectStore.SetValue	C#
 	0x8 in Uno.UI.Controls.BindableView.SetValue	C#
>	0x7 in Chinook.DataLoader.DataLoaderView.set_State at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderView.Properties.cs:55,19	C#
 	0x3 in Chinook.DataLoader.DataLoaderView.SetState at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderView.cs:81,4	C#
 	0x3 in Chinook.DataLoader.DataLoaderView.Chinook.DataLoader.IDataLoaderViewDelegate.SetState at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderView.cs:91,4	C#
 	0x28 in Chinook.DataLoader.DataLoaderViewControllerBase<Windows.UI.Core.CoreDispatcher>.UpdateUI at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderViewControllerBase.cs:275,4	C#
 	0x8 in Chinook.DataLoader.DataLoaderViewControllerBase<Windows.UI.Core.CoreDispatcher>.<Update>b__41_1 at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderViewControllerBase.cs:256,7	C#
 	0x4E in Chinook.DataLoader.DataLoaderViewController.RunOnDispatcher at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno\DataLoaderViewController.cs:20,5	C#
 	0x33 in System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<Chinook.DataLoader.DataLoaderViewController.<RunOnDispatcher>d__1> at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:316,17	C#
 	0x2E in Chinook.DataLoader.DataLoaderViewController.RunOnDispatcher	C#
 	0x56B in Chinook.DataLoader.DataLoaderViewControllerBase<Windows.UI.Core.CoreDispatcher>.Update at C:\Repos\Chinook.DataLoader\src\DataLoader.Uno.Shared\DataLoaderViewControllerBase.cs:254,6	C#
 	0x6 in System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.InvokeMoveNext at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1092,17	C#
 	0x73 in System.Threading.ExecutionContext.RunInternal at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:968,17	C#
 	0x4 in System.Threading.ExecutionContext.Run at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:910,13	C#
 	0x32 in System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1073,25	C#
 	0x6 in System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.<>c.<.cctor>b__7_0 at /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/external/corert/src/System.Private.CoreLib/src/System/Threading/Tasks/TaskContinuation.cs:379,78	C#
 	0xC in Uno.UI.Dispatching.CoreDispatcherSynchronizationContext. at C:\a\1\s\src\Uno.UI.Dispatching\Core\CoreDispatcherSynchronizationContext.cs:25,46	C#
 	0x63 in Uno.UI.Dispatching.CoreDispatcher.InvokeOperation at C:\a\1\s\src\Uno.UI.Dispatching\Core\CoreDispatcher.cs:379,5	C#
 	0x2 in Uno.UI.Dispatching.CoreDispatcher.InvokeOperationSafe at C:\a\1\s\src\Uno.UI.Dispatching\Core\CoreDispatcher.cs:343,5	C#
 	0xAB in Uno.UI.Dispatching.CoreDispatcher.DispatchItems at C:\a\1\s\src\Uno.UI.Dispatching\Core\CoreDispatcher.cs:305,6	C#
 	0x6 in Uno.UI.Dispatching.CoreDispatcher.CoreDispatcherImplementor.Java.Lang.IRunnable.Run at C:\a\1\s\src\Uno.UI.Dispatching\Core\CoreDispatcher.Android.cs:146,5	C#
 	0x8 in Java.Lang.IRunnableInvoker.n_Run at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-31/mcw/Java.Lang.IRunnable.cs:84,4	C#
 	0x8 in Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PP_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:22,5	C#
```

Here are the values in DataLoaderView (for when the call stack is focused at the marked position (with `>`).
![image](https://user-images.githubusercontent.com/39710855/232049718-c58892aa-bbb2-4420-9565-047a2eb41405.png)

The XAML has this hierarchy:
```
Page
  DataLoaderView
    DataTemplate (via ContentPresenter of the DataLoaderView's ControlTemplate)
      ContentControl bound on DataLoaderView.State.Data (This Data property would normally contain the ViewModel)
        DataTemplate (via DataTemplateSelector)
          ComboBox bound to the property (`SelectedItem="{Binding Property, Mode=TwoWay}"`          
````

**Comment by jeromelaban** [comment]
This is likely related to a two-way binding to the ComboBox, when the DataContext changes (probably to null). A potential workaround be to change the order of the properties on the ComboBox to put the `SelectedItem(s)` property first (or last, depending on the current XAML).

**Comment by jeanplevesque** [comment]
Interesting! I removed the `Mode=TwoWay` and the exception no longer happens.
I obviously can't use that as a fix, but that's interesting.
I'll try playing around with the order of the properties.

**Comment by jeanplevesque** [comment]
This is the original setup (which leads to the exception when resetting the DataContext).
```xml
<ComboBox x:Name="Bob"
	  ItemsSource="{Binding Items}"
          SelectedItem="{Binding Property, Mode=TwoWay}"
	  ItemTemplate="{StaticResource EnumComboBoxItemTemplate}"
	  IsEnabled="{Binding ItemsSource.Count, ElementName=Bob, Converter={StaticResource MoreThanOneToTrue}}"
	  Style="{StaticResource ComboBoxWithSeparatorStyle}" />
```

I tried the following instead (swapped `ItemsSource` and `SelectedItem`).
This fixes the exceptions, but breaks the initial value of the combobox. 🤔
```xml
<ComboBox x:Name="Bob"
          SelectedItem="{Binding Property, Mode=TwoWay}"
          ItemsSource="{Binding Items}"
	  ItemTemplate="{StaticResource EnumComboBoxItemTemplate}"
	  IsEnabled="{Binding ItemsSource.Count, ElementName=Bob, Converter={StaticResource MoreThanOneToTrue}}"
	  Style="{StaticResource ComboBoxWithSeparatorStyle}" />
```

Something to note is that the second setup is not suitable for UWP. The missing initial selection happens on all platforms, including UWP. That suggests that it's definitely not the correct setup.

With that in mind, I think the first setup is the correct one, as it doesn't have any problem on UWP.

**Comment by jeromelaban** [comment]
@jeanplevesque could you create a repro solution for this issue? Thanks!

**Comment by jeanplevesque** [comment]
[ComboBoxSample.zip](https://github.com/unoplatform/uno/files/12123509/ComboBoxSample.zip)

I managed to create a repro sample that highlights the difference between Windows and Mobile.
When you start the app, hit "Clear DataContext".
- WinUI: Nothing special happens.
- Android: ObjectDisposedException is thrown (because the property **getter** of the two-way binding is invoked, which isn't the case on WinUI).

**Comment by jeromelaban** [comment]
The changes for this issue are now available in 4.9.47 and 4.10.22.
